### DISCUSSION TITLE
1.10.0

### DISCUSSION THREAD

**Discussion by stephencelis** [pointfreeco/swift-snapshot-testing#642] [General]
## Breaking changes

- Changed: The package name has been changed from `SnapshotTesting` to `swift-snapshot-testing`, to better reflect community naming conventions (https://github.com/pointfreeco/swift-snapshot-testing/pull/555). This may require you to update any `Package.swift` files that depend on Snapshot Testing.
- Deprecated: support for CocoaPods and Carthage has been deprecated. Please use the Swift Package Manager to depend on Snapshot Testing.
- Deprecated: Snapshot Testing now requires Swift 5.5+, as well at iOS 13+, macOS 10.15+, tvOS 13+, and watchOS 6+.

> **Warning**: The following updates change how a strategy may output a snapshot, causing existing snapshot suites to fail. When upgrading, please rerecord your failing snapshots and audit the changes.

- Fixed: A UIViewController strategy was not passing along the correct trait collection (thanks @arnopoulos, https://github.com/pointfreeco/swift-snapshot-testing/pull/554).
- Fixed: iPhone XS Max portrait configuration was not applying base traits (thanks @imvm, https://github.com/pointfreeco/swift-snapshot-testing/pull/417)
- Fixed: URLRequest strategy now sorts query items for more consistent diffing (thanks @mihai8804858, https://github.com/pointfreeco/swift-snapshot-testing/pull/491).
- Removed: GLKView is no longer supported (thanks @jpsim, https://github.com/pointfreeco/swift-snapshot-testing/pull/507).

## Everything else

- Added: Image-based strategies have a new `perceptualPrecision` option, which can be used to support snapshot tests across Intel and M1 devices. Along for the ride is a 90+% speed improvement to the strategies (thanks @ejensen, https://github.com/pointfreeco/swift-snapshot-testing/pull/628).
- Added: Windows support, CI (thanks @MaxDesiatov, https://github.com/pointfreeco/swift-snapshot-testing/pull/532).
- Added: Failure messages now include the snapshot name, if given (thanks @calda, https://github.com/pointfreeco/swift-snapshot-testing/pull/547).
- Added: iPhone 13 configs (thanks @luispadron, https://github.com/pointfreeco/swift-snapshot-testing/pull/603).
- Added: iPad 9.7" and iPad 10.2" configs (thanks @skols85, https://github.com/pointfreeco/swift-snapshot-testing/pull/405).
- Added: iPhone 12 and iPhone 12 Pro Max configs(thanks @imvm, https://github.com/pointfreeco/swift-snapshot-testing/pull/418).
- Added: `Snapshotting.json` for `Any` value (thanks @NachoSoto, https://github.com/pointfreeco/swift-snapshot-testing/pull/552).
- Added: Newly-recorded snapshots are now attached to the test run (thanks @marcelofabri, https://github.com/pointfreeco/swift-snapshot-testing/pull/586).
- Fixed: the UIImage strategy was using the device scale instead of the scale of the given images (thanks @codeman9, https://github.com/pointfreeco/swift-snapshot-testing/pull/472).
- Fixed: the UIImage strategy now compares image contexts using the same colorspace (thanks @dflems, https://github.com/pointfreeco/swift-snapshot-testing/pull/446).
- Fixed: the WebView strategy no longer overrides the delegate (thanks @teameh in https://github.com/pointfreeco/swift-snapshot-testing/pull/443).
- Fixed: Patched a leak when running tests in a host application (thanks @llinardos, https://github.com/pointfreeco/swift-snapshot-testing/pull/511).
- Fixed: False positive when asserting reference image with empty image (thanks @nuno-vieira, https://github.com/pointfreeco/swift-snapshot-testing/pull/453).
- Fixed: watchOS compilation (thanks @aydegee, https://github.com/pointfreeco/swift-snapshot-testing/pull/579).
- Fixed: Support for repeatedly running tests (thanks @krzysztofpawski, https://github.com/pointfreeco/swift-snapshot-testing/pull/585).
- Infrastructure: Add `swift-snapshot-testing-stitch` to plugins list (thanks @Sherlouk, https://github.com/pointfreeco/swift-snapshot-testing/pull/483)
- Infrastructure: Document diff tool configuration.
- Infrastructure: README updates (thanks @MaatheusGois, @gohanlon), documentation fixes (thanks @heckj, @valeriyvan).
- Infrastructure: Added GitHub issue templates (https://github.com/pointfreeco/swift-snapshot-testing/pull/556).
- Infrastructure: Add SnapshotTestingHEIC to plugins list (thanks @alexey1312, https://github.com/pointfreeco/swift-snapshot-testing/pull/561).

## New Contributors
* @arietis made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/468
* @codeman9 made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/472
* @Nikoloutsos made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/482
* @mihai8804858 made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/491
* @dflems made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/446
* @skols85 made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/405
* @imvm made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/418
* @teameh made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/443
* @MaatheusGois made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/496
* @calda made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/547
* @MaxDesiatov made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/532
* @jpsim made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/507
* @arnopoulos made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/554
* @llinardos made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/511
* @heckj made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/601
* @aydegee made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/579
* @valeriyvan made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/630
* @ejensen made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/628
* @luispadron made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/603
* @alexey1312 made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/561
* @gohanlon made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/592
* @marcelofabri made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/586
* @nuno-vieira made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/453
* @krzysztofpawski made their first contribution in https://github.com/pointfreeco/swift-snapshot-testing/pull/585

**Full Changelog**: https://github.com/pointfreeco/swift-snapshot-testing/compare/1.9.0...1.10.0

<hr /><em>This discussion was created from the release <a href='https://github.com/pointfreeco/swift-snapshot-testing/releases/tag/1.10.0'>1.10.0</a>.</em>

**Comment by MagFer** [comment]
Good afternoon @stephencelis, is there a chance to bring back support to Carthage or create a precompiled XCFramework for every release? 

I see there are other people interested in this feature: https://github.com/pointfreeco/swift-snapshot-testing/discussions/648.
FYI @lazar-otasevic-cif

**Reply by stephencelis** [reply]
We do not intend to bring back Carthage or CocoaPods support. See https://github.com/pointfreeco/swift-snapshot-testing/issues/644 for more of the rationale behind the decision.

That said, we believe that anyone that wants to use Carthage or CocoaPods is free to do so, but they must maintain their own project or podfile.

As for using a precompiled `.xcframework` file. We have several users that use `.xcframework`s of our [swift-composable-architecture](https://github.com/pointfreeco/swift-composable-architecture) library, which has never had Carthage support. Users that have wanted an `.xcframework` have managed that build process themselves.

**Reply by NachoSoto** [reply]
It is unfortunate that this breaking change was released in a ["minor" release](https://semver.org) though. I had to fix a bunch of repos that got this update automatically and no longer built.

**Reply by stephencelis** [reply]
@NachoSoto Sorry about that! I know that switching from CocoaPods/Carthage is not trivial, and neither is maintaining your own Podfile or Xcode project file for Carthage support, but just to understand, the act of releasing this new version didn't break your existing projects, right? It's more that you had to do some work to upgrade?

While we understand that there's some friction in removing official support for these distribution mechanisms, we aim to keep our SemVer versioning based off source compatibility, not distribution management: https://github.com/pointfreeco/swift-snapshot-testing/discussions/647#discussioncomment-3736262

**Reply by NachoSoto** [reply]
> but just to understand, the act of releasing this new version didn't break your existing projects, right? It's more that you had to do some work to upgrade?

It did. Not a huge deal, but I had to change the version specifier from `1.x` to hardcode `1.9.0` (both in projects using CocoaPods and Carthage).

> While we understand that there's some friction in removing official support for these distribution mechanisms, we aim to keep our SemVer versioning based off source compatibility, not distribution management

I see. But unfortunately that decision meant that overnight several projects I work on stopped working until I made that change.

**Reply by stephencelis** [reply]
Hm, I'm actually surprised that CocoaPods would break since the specs are hosted by them, though I do understand how Carthage would break. Sorry again for the breakage. It's a learning experience for us, but a little late for us to back out of now.

**Reply by MagFer** [reply]
Good afternoon @stephencelis. 

We've done the same as @NachoSoto, and fixed the framework version to 1.9.0 in the Cartfile. There's not a significant inconvenience moving from CocoaPods to SPM. However, the precompiled XCFrameworks from Carthage make life easier.

Do you have any advice on how to generate a v1.10.0 XCFramework? Could we run the following command?
`carthage build --no-skip-current`

**Comment by balavor** [comment]
`Carthage` still works better for large projects since SPM is not able to pre-build dependencies in cross-platform `xcframeworks`. And it doesn't fetch the entire repo from the origin...

**Comment by MagFer** [comment]
Hi all! 

Do we have news on how to use a newer `swift-snapshot-testing` version than 1.9.0 in Carthage or download the XCFramework?

We are still using `v1.9.0` and I don't think we can precompile a Swift package using the next command, since the library repo doesn't have any Xcode project file.

```
xcrun xcodebuild archive \
  -workspace . \
  -scheme {scheme} \
  -configuration Release \
  -destination 'generic/platform=iOS' \
  -derivedDataPath '.build' \
  -archivePath './.build/Release/{library}-iphoneos.xcarchive' \
  -allowProvisioningUpdates \
  SKIP_INSTALL=NO \
  BUILD_LIBRARY_FOR_DISTRIBUTION=YES
```

I haven't found posts on how to precompile a Swift Package either.
https://forums.swift.org/t/how-to-build-swift-package-as-xcframework/41414/51
